### ISSUE TITLE
[C++] Create asynchronous version of StreamReader

### ISSUE THREAD

**Issue by toddfarmer** [toddfarmer/arrow-migration#1015]
***Note**: This issue was originally created as [ARROW-1009](https://issues.apache.org/jira/browse/ARROW-1009). Please see the [migration documentation](https://gist.github.com/toddfarmer/12aa88361532d21902818a6044fda4c3) for further details.*

### Original Issue Description:
the `AsyncStreamReader` would buffer the next record batch in a background thread, while emulating the current synchronous / blocking API

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1009?focusedCommentId=16761428) by Wes McKinney (wesm):*
It might be nice to solve this in the context of gRPC / Flight, to define a general async stream reader API

cc `[~pitrou]`

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1009?focusedCommentId=17200883) by Antoine Pitrou (apitrou):*
We have `StreamDecoder` now, which is an event-driven IPC reader.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1009?focusedCommentId=17287351) by Wes McKinney (wesm):*
What do you think about async APIs in RecordBatchReader? cc `[~lidavidm]` `[~westonpace]`

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1009?focusedCommentId=17287399) by David Li (lidavidm):*
I think the main question is a use case for them. Right now for instance Flight uses only blocking gRPC APIs so say if we wanted to do compression asynchronously, even if there were an asynchronous reader available, we'd still have to explicitly manage background threads and queues at some place. (And I'm not sure if the asynchronous gRPC APIs adapt easily to what we're doing/it would be a decent amount of work to wire the async gRPC APIs to our async API.)

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1009?focusedCommentId=17288568) by Weston Pace (westonpace):*
I'm a little confused by the ticket I'm afraid.  The only `StreamReader` I see is `parquet/stream_reader.h` and that is focused on C++ `>>` streams.  It allows the user to read elements out of a parquet file one at a time (presumably into some non-arrow format?)

There is also the `RecordBatchStreamReader` which allows for reading a source one record batch at a time (instead of an entire table at a time).  An asynchronous mirror version of this API would make sense in the same way it would make sense to have such an API for any I/O operation.  The implementation would probably come after an asynchronous table reader for the given format (e.g. once an asynchronous table reader is done for parquet then one can tackle an asynchronous stream reader).

There is StreamDecoder which allows processing buffers through observation.  It would have similar memory pressure to RecordBatchStreamReader if I understand because the stream decoder will buffer up to a record batch and then deliver it.  It's also synchronous and there could be an asynchronous version (i.e. `ipc::StreamDecoder::ConsumeAsync`).

Then there is flight mentioned in the comments.  Flight uses `MetadataRecordBatchReader`.  It could be made asynchronous.

For the flight client/server I agree that there might not be a strong motivation right now.  Since it isn't yet used internally by Arrow (in the same way the query/dataset APIs use the various readers internally) it isn't as needed.  I think it could still be a low priority improvement to keep I/O on the IOContext and CPU on the CPU thread pool.  Any consumers can do this themselves but it would be a "nice to have" if Arrow took care of the queuing for them.

Do we think a flight server might serve up hundreds/thousands of concurrent downloads?  For example, you might have this in video hosting (e.g. YouTube) but there is a natural throttling in place there (no need to send data faster than it is being displayed) that might not apply for data analytics.

`[~wesm]` Can you provide a few more details what the intent is for this issue?  I think the highest priority of all the things I mentioned is getting an asynchronous record batch reader from parquet (will be needed for dataset nested parallelism) and I think we are close given the prebuffering is already pushing I/O to the IOContext.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1009?focusedCommentId=17356978) by Wes McKinney (wesm):*
The intent was to provide an asynchronous API for `RecordBatchReader`. So this might be better written as `AsyncRecordBatchReader`

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1009?focusedCommentId=17358853) by Weston Pace (westonpace):*
I think this is completed at this point.   There are asynchronous versions of the various record batch readers.  There is not an AsyncRecordBatchReader though.  RecordBatchReader is basically Iterator<RecordBatch> + GetMetadata().  There has not been the need for a similar interface on the async side.  Generally the schema is either passed along with the reader or it is grabbed from the FileReader.  I'm not sure we need to do anything here any more.  `[~lidavidm]` thoughts?

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1009?focusedCommentId=17358855) by David Li (lidavidm):*
I think we can add the interface when warranted. For instance, if/when we ever integrate async APIs into Flight, which has various reader classes for which a common interface would at least ensure consistency.
